### PR TITLE
Add solution for TASK-3

### PR DESCRIPTION
**Pull request by perkeje**

### CODE REVIEW

File: src/main/java/net/barrage/school/java/ecatalog/app/ProductServiceImpl.java [hunk 1]
```diff
@@ -2,13 +2,20 @@
...
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
 import org.springframework.stereotype.Service;
 
 import java.util.ArrayList;
 import java.util.Comparator;
 import java.util.List;
 
 @Service
+@EnableCaching
```
Comment (AntonRomankovSetronica) [inline src/main/java/net/barrage/school/java/ecatalog/app/ProductServiceImpl.java]: All `Enable*` like annotations are only suitable for configrations. U cant just put it to random bean. https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/cache/annotation/EnableCaching.html

File: src/main/java/net/barrage/school/java/ecatalog/app/ProductServiceImpl.java [hunk 2]
```diff
@@ -28,8 +36,16 @@ public List<Product> listProducts() {
...
     }
 
+    @Scheduled(fixedDelayString = "${ecatalog.products.fetchDelay}")
+    @CacheEvict(value = "products", allEntries = true)
+    public void clearProductCache() {
+    }
+
     @Override
     public List<Product> searchProducts(String query) {
+        // I read something about cache not working if called from method from same bean, so this will never call
```
Comment (AntonRomankovSetronica) [inline src/main/java/net/barrage/school/java/ecatalog/app/ProductServiceImpl.java]: Yeap! That's great u've debug it)

There is a trick for such case - self inject
https://www.baeldung.com/spring-self-injection

File: src/test/java/net/barrage/school/java/ecatalog/app/ProductServiceImplTest.java [hunk 3]
```diff
@@ -18,4 +19,14 @@ class ProductServiceImplTest {
...
     }
+
+    @Test
+    void product_search_is_not_empty() {
+        assertFalse(impl.searchProducts("Socks smell better").isEmpty(), "Expect listProducts() to return something.");
+    }
+
+    @Test
+    void product_search_is_empty() {
+        assertTrue(impl.searchProducts("THIS IS EMPTY MESSAGE").isEmpty(), "Expect listProducts() to return something.");
```
Comment (AntonRomankovSetronica) [inline src/test/java/net/barrage/school/java/ecatalog/app/ProductServiceImplTest.java]: wrong error msg

File: src/main/java/net/barrage/school/java/ecatalog/app/ProductServiceImpl.java (hunk 1 above)
Comment (perkeje) [reply src/main/java/net/barrage/school/java/ecatalog/app/ProductServiceImpl.java]: I have put it to ECatalogApplication is this okay?

File: src/main/java/net/barrage/school/java/ecatalog/app/ProductServiceImpl.java (hunk 2 above)
Comment (perkeje) [reply src/main/java/net/barrage/school/java/ecatalog/app/ProductServiceImpl.java]: Thanks. Should I leave it for now or implement self inject?

**Review by AntonRomankovSetronica — APPROVED** [review]

**Pull request merged**